prism/src/main/java/org/apache/falcon/resource/AbstractEntityManager.java

```diff
@@ -282,9 +289,56 @@ public synchronized APIResult update(HttpServletRequest request, String type, St
             throw FalconWebException.newException(e, Response.Status.BAD_REQUEST);
         } finally {
             ConfigurationStore.get().cleanupUpdateInit();
+            releaseUpdateEntityLocks(entityName, tokenList);
         }
     }
 
+    private List<MemoryLocks.LockToken> obtainUpdateEntityLocks(Entity entity)
```

Comment:
The design of obtaining fine-grained object level locks is very easily prone to deadlocks. E.g. consider the scenario when 2 threads are trying to obtain locks for entities in reverse order: {e1, e2} and {e2, e1}. Such cases can be extremely hard to debug.

This strategy can only be taken if 1) a particular sequence of locks is agreed upon in the entire code base. 2) All lock attempts are only made in forward direction.